cpp/src/arrow/util/async_generator.h

```diff
@@ -1001,32 +1004,45 @@ class MergedGenerator {
   };
 
   struct InnerCallback {
```

Comment:
Ok, so let me make sure I understand the change.  If a merged generator pulled off a generator and that generator had a whole bunch of already completed futures (e.g. vector generator) then we would end up recursing for each call into the child.

Instead, you setup the TryAddCallback loop to only recurse when the future isn't finished.  I think this is a great addition.  Even if we weren't overflowing the stack in most cases we'd still be creating unnecessarily long stack traces.  Thanks for working through this.

Reply:
Yup, that's it. And in particular the degenerate case here is that the synchronously-run callback from the mapped generator's futures would pull from the mapped generator, creating a loop implemented as (non-tail) recursion.